src/plots/geo/geo2.js

```diff
@@ -0,0 +1,612 @@
+
+    var choroplethLayer = this.layers.backplot.select('.choroplethlayer');
+    this.dataPaths.choropleth = choroplethLayer.selectAll('path');
+
+    this.updateFx(fullLayout, geoLayout);
+    this.render();
+};
+
+proto.updateProjection = function(fullLayout, geoLayout) {
+    var gs = fullLayout._size;
+    var domain = geoLayout.domain;
+    var projLayout = geoLayout.projection;
+    var rotation = projLayout.rotation || {};
+    var center = geoLayout.center || {};
+
+    var projection = this.projection = getProjection(geoLayout);
+
+    // set 'pre-fit' projection
+    projection
+        .center([center.lon - rotation.lon, center.lat - rotation.lat])
+        .rotate([-rotation.lon, -rotation.lat, rotation.roll])
+        .parallels(projLayout.parallels);
+
+    // setup subplot extent [[x0,y0], [x1,y1]]
+    var extent = [[
+        gs.l + gs.w * domain.x[0],
+        gs.t + gs.h * (1 - domain.y[1])
+    ], [
+        gs.l + gs.w * domain.x[1],
+        gs.t + gs.h * (1 - domain.y[0])
+    ]];
+
+    var lonaxis = geoLayout.lonaxis;
+    var lataxis = geoLayout.lataxis;
+    var rangeBox = makeRangeBox(lonaxis.range, lataxis.range);
+
+    // fit projection 'scale' and 'translate' to set lon/lat ranges
+    projection.fitExtent(extent, rangeBox);
+
+    var b = this.bounds = projection.getBounds(rangeBox);
+    var s = this.fitScale = projection.scale();
+    var t = projection.translate();
+
+    if(
+        !isFinite(b[0][0]) || !isFinite(b[0][1]) ||
+        !isFinite(b[1][0]) || !isFinite(b[1][1]) ||
+        isNaN(t[0]) || isNaN(t[0])
+    ) {
+        Lib.warn('Invalid geo settings');
+
+        // TODO fallback to default ???
+    }
+
+    // px coordinates of view mid-point,
+    // useful to update `geo.center` after interactions
+    var midPt = this.midPt = [
+        b[0][0] + (b[1][0] - b[0][0]) / 2,
+        b[0][1] + (b[1][1] - b[0][1]) / 2
+    ];
+
+    // adjust projection to user setting
+    projection
+        .scale(projLayout.scale * s)
+        .translate([t[0] + (midPt[0] - t[0]), t[1] + (midPt[1] - t[1])])
+        .clipExtent(b);
+
+    // the 'albers usa' projection does not expose a 'center' method
+    // so here's this hack to make it respond to 'geoLayout.center'
+    if(geoLayout._isAlbersUsa) {
+        var centerPx = projection([center.lon, center.lat]);
+        var tt = projection.translate();
+
+        projection.translate([
+            tt[0] - (centerPx[0] - tt[0]),
+            tt[1] - (centerPx[1] - tt[1])
+        ]);
+    }
+};
+
+proto.updateBaseLayers = function(fullLayout, geoLayout) {
+    var _this = this;
+    var topojson = _this.topojson;
+    var layers = _this.layers;
+    var basePaths = _this.basePaths;
+
+    function isAxisLayer(d) {
+        return (d === 'lonaxis' || d === 'lataxis');
+    }
+
+    function isTopoLayer(d) {
+        return (
+            constants.fillLayers.indexOf(d) !== -1 ||
+            constants.lineLayers.indexOf(d) !== -1
```

Comment:
🐎  is it worth turning this (and `isAxisLayer`) into a hash lookup `{ocean: 1, land: 1, lakes: 1}` instead of `indexOf`? in fact it looks like `indexOf` is the only way `fillLayers` and `lineLayers` are ever used

Reply:
Very good idea :racehorse:

Reply:
done in https://github.com/plotly/plotly.js/pull/2030/commits/eeee9118294af10660a7bcedfbbcea12ec275579